Inventory transaction resources

cookeac opened the pull request:
Defines a generalised icarInventoryTransactionResource and its specific versions, icarFeedTransactionResource and icarMedicineTransactionResource. Resolves #369

Note - update to exampleURLScheme included, but I know that I will need to merge these changes with the changes in #381 before this can be merged.

cookeac committed:
Inventory transaction resources

Defines a generalised icarInventoryTransactionResource and its specific versions, icarFeedTransactionResource and icarMedicineTransactionResource.
Resolves #369

cookeac committed:
Fix linting errors

cookeac committed:
Fix linting errors

Remove $id terms

cookeac committed:
Fix linting error

Removed extraneous comma

AndreasSchultzGEA commented on types/icarInventoryTransactionType.json:
```diff
@@ -0,0 +1,58 @@
+                    "type": "number",
+                    "description": "The overall volume, weight or count of the product in the transaction in the units defined."
+                },
+                "units": {
```
Singular instead of plural?

cookeac replied:
In English we would almost always use plural if the value/amount measure could be greater than one. An alternative is unitCode, which is more descriptive but not consistent with other places (e.g. icarMassMeasureType.json).

AndreasSchultzGEA commented on types/icarProductReferenceType.json:
```diff
@@ -0,0 +1,44 @@
+            ],
+
+            "properties": {
+                "identifiers": {
```
from icarResourceReferenceType, there is the included identifier-property of type icarIdentifierType. 
Shouldn't the array better be named alternativeIdentifier as everywhere else? 
And: The type of identifier and the array of alternativeIds differ (icarProductIdentifierType).
Ok, the content is the same. But still different types.

cookeac replied:
Noted that we do use `alternativeIdentifiers` in locations and animals. In those cases there tends to be one formal identifier and several alternatives. I'm less convinced with products, medicines, etc where there seem to be a range of identifiers. The source system has obviously chosen one to be its `identifier`, but presumably the destination may use another. Happy to change this to `alternativeIdentifiers` for consistency, but not sure it is needed.

Noted on the types - but that is a feature of the base `icarResourceReferenceType` class and was already the case with medicines, for instance.

AndreasSchultzGEA commented on types/icarMedicineReferenceType.json:
```diff
@@ -2,15 +2,12 @@
     "allOf": [
         {
-            "$ref": "icarResourceReferenceType.json"
+            "$ref": "../types/icarProductReferenceType.json"
```
may be a breaking change. But only new properties are added by the changes inheritance. Unsure!

erwinspeybroeck replied:
Andrew, a general remark from my side.
You have a generic API for inventory-transactions: GET /locations/{location-scheme}/{location-id}/inventory-transactions and specific ones for feed and medicines. But in the specific ones also the complete list of family products is taken _[ Animal Feeds, Animal Reproductive Products, Veterinary Supplies, Seed and Plant Material, Fertilisers and Nutrients, Pest Control Products, Other Animal Products, Milking Supplies, Fencing Supplies, Water System Supplies ]_, while one would expect this to be limited to e.g. Animal Feeds in case of feed. 
Not a blocking remark, I don't know if we can have a fixed family ...

cookeac replied:
As you say, only new properties are added, and when parsed by code generation the result should be consistent. I also ensured that the new properties are not required.

erwinspeybroeck commented:
Andrew, a general remark from my side.
You have a generic API for inventory-transactions: GET /locations/{location-scheme}/{location-id}/inventory-transactions and specific ones for feed and medicines. But in the specific ones also the complete list of family products is taken _[ Animal Feeds, Animal Reproductive Products, Veterinary Supplies, Seed and Plant Material, Fertilisers and Nutrients, Pest Control Products, Other Animal Products, Milking Supplies, Fencing Supplies, Water System Supplies ]_, while one would expect this to be limited to e.g. Animal Feeds in case of feed. 
Not a blocking remark, I don't know if we can have a fixed family ...

cookeac commented:
> Andrew, a general remark from my side. You have a generic API for inventory-transactions: GET /locations/{location-scheme}/{location-id}/inventory-transactions and specific ones for feed and medicines. But in the specific ones also the complete list of family products is taken _[ Animal Feeds, Animal Reproductive Products, Veterinary Supplies, Seed and Plant Material, Fertilisers and Nutrients, Pest Control Products, Other Animal Products, Milking Supplies, Fencing Supplies, Water System Supplies ]_, while one would expect this to be limited to e.g. Animal Feeds in case of feed. Not a blocking remark, I don't know if we can have a fixed family ...

You're exactly right, and I was aware of this. The only thing I could think was documentation - that if you were using the feed or medicine resource type, the family should be set to a constant value. I'll ask my team if there is any way to use the "default" keyword if the property is already defined.

AndreasSchultzGEA reviewed (commented):
Hello Andrew, I'm sorry but I couldn't finish the review due to missing time.

cookeac committed:
Merge branch 'adewg:Develop' into Develop

erwinspeybroeck commented on types/icarFeedReferenceType.json:
a bit strange to see the url with Pure-Farming in here

cookeac replied:
Ah, I didn't catch making that one relative. Will fix.

cookeac commented on types/icarFeedReferenceType.json:
```diff
@@ -0,0 +1,22 @@
+            "type": "object",
+            "properties": {
+                "category": {
+                    "$ref": "https://raw.githubusercontent.com/Pure-Farming/ICAR/ADE-1/enums/icarFeedCategoryType.json",
```
```suggestion
                    "$ref": "../enums/icarFeedCategoryType.json",
```

cookeac commented on types/icarFeedReferenceType.json:
```diff
@@ -0,0 +1,22 @@
+                    "description": "Defines the category of the feed product."
+                },
+                "type": {
+                    "$ref": "https://raw.githubusercontent.com/Pure-Farming/ICAR/ADE-1/types/icarFeedIdentifierType.json",
```
```suggestion
                    "$ref": "../types/icarFeedIdentifierType.json",
```

cookeac committed:
Correct paths

cookeac committed:
Correct path in types/icarFeedReferenceType.json

cookeac committed:
Merge branch 'adewg:Develop' into Develop

cookeac committed:
Merge branch 'adewg:Develop' into Develop

cookeac committed:
Add inventory transactions to URL schemes

Inventory transactions are already in the example URL scheme. Adding the medicines ones to the health scheme, and the feed ones to the feed scheme.

cookeac commented:
Reviewed at previous meetings; merging.
